python/lsst/daf/butler/core/registry.py

```diff
@@ -189,6 +189,12 @@ def __init__(self, registryConfig, schemaConfig=None, dimensionConfig=None, crea
         for packerFactory in self._dataIdPackerFactories.values():
             packerFactory.updateFieldsToGet(self._fieldsToAlwaysGet)
 
+    @abstractmethod
+    def shutdown(self):
```

Comment:
The short answer is no, I tried that first as I agree that is better. Unfortunately (I think because something somewhere still has references, or perhaps the order in which they are cleaned up) this still causes an exception to be raised when sqlalchemy does its own object clean up. The only way I could get it to not raise was explicitly calling this in whatever owned the registry.

Reply:
It may be possible to reason and trace through the destructor orders, but I would prefer to get that done in another ticket, as it may turn into a decent chunk of work itself.

Reply:
Python doesn't guarantee that `__del__` will be called so it's effectively useless. The usual recommendation is to provide `close`/`shutdown` methods and context managers.